libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/target/ParentChildrenSyncDownTarget.kt

```diff
@@ -0,0 +1,363 @@
+                childrenInfo.idFieldName
+            )
+        }
+        return localIdsSize
+    }
+
+    @Throws(IOException::class, JSONException::class)
+    protected fun getChildrenRemoteIdsWithSoql(
+        syncManager: SyncManager,
+        soqlForChildrenRemoteIds: String
+    ): Set<String> {
+
+        // Makes network request and parses the response.
+        var records = startFetch(syncManager, soqlForChildrenRemoteIds)
+        val remoteChildrenIds: MutableSet<String> = HashSet(parseChildrenIdsFromResponse(records))
+        while (true) {
+            syncManager.checkAcceptingSyncs()
+            // Fetch next records, if any.
+            records = continueFetch(syncManager) ?: break
+            remoteChildrenIds.addAll(parseIdsFromResponse(records))
+        }
+        return remoteChildrenIds
+    }
+
+    protected fun parseChildrenIdsFromResponse(records: JSONArray): Set<String> {
+        return with(HashSet<String>()) {
+            JSONObjectHelper
+                .toList<JSONObject>(records)
+                .forEach { record ->
+                    val childrenRecords = record.optJSONArray(childrenInfo.sobjectTypePlural) ?: JSONArray()
+                    this.addAll(parseIdsFromResponse(childrenRecords))
+                }
+            this
+        }
+    }
+
+    override fun getQuery(maxTimeStamp: Long): String {
+        val childrenWhere = StringBuilder()
+        val parentWhere = StringBuilder()
+        if (maxTimeStamp > 0) {
+            // This is for re-sync
+            //
+            // Ideally we should target parent-children 'groups' where the parent changed or a child changed
+            //
+            // But that is not possible with SOQL:
+            //   select fields, (select childrenFields from children where lastModifiedDate > xxx)
+            //   from parent
+            //   where lastModifiedDate > xxx
+            //   or Id in (select parent-id from children where lastModifiedDate > xxx)
+            // Gives the following error: semi join sub-selects are not allowed with the 'OR' operator
+            //
+            // Also if we do:
+            //   select fields, (select childrenFields from children where lastModifiedDate > xxx)
+            //   from parent
+            //   where Id in (select parent-id from children where lastModifiedDate > xxx or parent.lastModifiedDate > xxx)
+            // Then we miss parents without children
+            //
+            // So we target parent-children 'goups' where the parent changed
+            // And we only download the changed children
+            childrenWhere.append(
+                buildModificationDateFilter(
+                    childrenInfo.modificationDateFieldName,
+                    maxTimeStamp
+                )
+            )
+            parentWhere.append(buildModificationDateFilter(modificationDateFieldName, maxTimeStamp))
+                .append(if (TextUtils.isEmpty(parentSoqlFilter)) "" else " and ")
```

Comment:
Optional: Another quick Kotlin update can be moving away from `TextUtils` in a lot of cases.  This could be `parentSoqlFilter.isNullOrEmpty()`.